web/modules/weather_data/src/Service/DataLayer.php

```diff
@@ -0,0 +1,245 @@
+            )->properties;
+        }
+        return $this->i_hourlyForecast;
+    }
+
+    public function getCurrentObservation($station)
+    {
+        // Don't stash this one away in an instance variable because it may need
+        // to be called with multiple stations in order to find a valid result.
+        // This is currently only called by getCurrentConditionsFromGrid, so it
+        // shouldn't duplicate any calls.
+        return $this->getFromWeatherAPI(
+            "/stations/$station/observations?limit=1",
+        )->features[0]->properties;
+    }
+
+    private $i_obsStations = false;
+    public function getObservationStations($wfo, $x, $y)
+    {
+        $wfo = strtoupper($wfo);
+        if (!$this->i_obsStations) {
+            $this->i_obsStations = $this->getFromWeatherAPI(
+                "/gridpoints/$wfo/$x,$y/stations",
+            )->features;
+        }
+        return $this->i_obsStations;
+    }
+
+    private $i_point = false;
+    public function getPoint($lat, $lon)
+    {
+        if (!$this->i_point) {
+            $lat = round($lat, 4);
+            $lon = round($lon, 4);
+
+            $this->i_point = $this->getFromWeatherAPI("/points/$lat,$lon");
+        }
+        return $this->i_point;
+    }
+
+    private function getPlaceNear($wktGeometry)
+    {
+        $sql = "SELECT
+          name,state,stateName,county,timezone,stateFIPS,countyFIPS
+          FROM weathergov_geo_places
+          ORDER BY ST_DISTANCE(point,ST_GEOMFROMTEXT('$wktGeometry'))
+          LIMIT 1";
+
+        $place = $this->database->query($sql)->fetch();
+
+        $place = (object) [
+            "city" => $place->name,
+            "state" => $place->state,
+            "stateName" => $place->stateName,
+            "stateFIPS" => $place->stateFIPS,
+            "county" => $place->county,
+            "countyFIPS" => $place->countyFIPS,
+            "timezone" => $place->timezone,
+        ];
+
+        return $place;
+    }
+
+    private $i_placeNearPoint = [];
+    public function getPlaceNearPoint($lat, $lon)
+    {
+        $key = "$lat $lon";
+        if (!$this->i_placeNearPoint[$key]) {
+            $this->i_placeNearPoint[$key] = $this->getPlaceNear(
+                "POINT($lon $lat)",
+            );
+        }
+        return $this->i_placeNearPoint[$key];
+    }
+
+    private $i_placeNearPolygon = [];
+    public function getPlaceNearPolygon($points)
+    {
+        $wktPoints = array_map(function ($point) {
+            return $point[0] . " " . $point[1];
+        }, $points);
+        $wktPoints = implode(",", $wktPoints);
+
+        if (!$this->i_placeNearPolygon[$wktPoints]) {
+            $this->iPlaceNearPolygon[$wktPoints] = $this->getPlaceNear(
+                "POLYGON(($wktPoints))",
+            );
+        }
+        return $this->iPlaceNearPolygon[$wktPoints];
+    }
+
+    public function databaseFetch($sql)
+    {
+        return $this->database->query($sql)->fetch();
```

Comment:
Thinking aloud: where do we want to catch and handle DB/query errors, should they emerge? I think in this state of the code, a given block `build()` method is the place that would catch these for now (and that might be the appropriate place)

Reply:
Added #871 so this question doesn't have to just live in your head. 😄